libbpf-tools/core_fixes.bpf.h

```diff
@@ -109,4 +109,39 @@ static __always_inline struct gendisk *get_disk(void *request)
 	return BPF_CORE_READ(r, q, disk);
 }
 
+/**
+ * commit 6521f8917082("namei: prepare for idmapped mounts") add `struct
+ * user_namespace *mnt_userns` as vfs_create() and vfs_unlink() first argument.
+ * At the same time, struct renamedata {} add `struct user_namespace
+ * *old_mnt_userns` item. Now, to kprobe vfs_create()/vfs_unlink() in a CO-RE
+ * way, determine whether there is a `old_mnt_userns` field for `struct
+ * renamedata` to decide which input parameter of the vfs_create() to use as
+ * `dentry`.
+ * see:
+ *     https://github.com/torvalds/linux/commit/6521f8917082
+ */
+struct renamedata___x {
+	struct user_namespace *old_mnt_userns;
+} __attribute__((preserve_access_index));
+
+static __always_inline bool renamedata_has_old_mnt_userns(void *request)
+{
+	struct renamedata___x *r = request;
+
+	if (bpf_core_field_exists(r->old_mnt_userns))
+		return true;
+	return false;
+}
+
+static __always_inline bool vfs_create_has_userns_argument(void)
+{
+	struct renamedata *tmp = NULL;
+	return renamedata_has_old_mnt_userns(tmp);
+}
+
+static __always_inline bool vfs_unlink_has_userns_argument(void)
+{
+	return vfs_create_has_userns_argument();
+}
+
```

Comment:
This is way too complicated, see https://github.com/OpenCloudOS/bpf-tools/pull/1.

Reply:
Yep, you are right.I'm wondering, the commit https://github.com/OpenCloudOS/bpf-tools/pull/1 the CO-RE suffix is `struct renamedata___v512`, does `6521f8917082` should be `v511`? I'm not sure `git describe` is correct or not.

```
linux]$ git describe 6521f8917082
v5.11-rc4-14-g6521f8917082
```

Reply:
I think we need only one of these functions.

Reply:
@chenhengqi Sorry, I modified it before as you suggested, but I overwrote that change last time I forced commit. Already submitted again, thanks.